zonec.c

```diff
@@ -1215,6 +1647,121 @@ zadd_rdata_txt_clean_wireformat()
+
+	memset(paramkeys, 0, sizeof(paramkeys));
+	/* 
+	 * In draft-ietf-dnsop-svcb-https-04 Section 7:
+	 * In wire format, the keys are represented by their numeric values in
+	 * network byte order, concatenated in ascending order.
+	 */
+	qsort( (void *)&parser->current_rr.rdatas[2]
+	     , parser->current_rr.rdata_count - 2
+	     , sizeof(rdata_atom_type *)
+	     , svcparam_key_cmp
+	     );
+
+	for (i = 2; i < parser->current_rr.rdata_count; i++) {
+		key  = read_uint16(rdata_atom_data(parser->current_rr.rdatas[i]));
+
+		/* In draft-ietf-dnsop-svcb-https-04 Section 7:
+		 *
+		 *     Keys (...) MUST NOT appear more than once.
+		 * 
+		 * If they key has already been seen, we have a duplicate
+		 */
+		if (!paramkeys[key])
+			/* keep track of keys that are present */
+			paramkeys[key] = 1;
+
+		else if (key < SVCPARAMKEY_COUNT) {
+			if(zone_is_slave(parser->current_zone->opts))
+				zc_warning_prev_line(
+					"Duplicate key found: %s\n",
+					svcparamkey_strs[key]);
+			else {
+				zc_error_prev_line(
+					"Duplicate key found: %s\n",
+					svcparamkey_strs[key]);
+			}
+		} else if(zone_is_slave(parser->current_zone->opts))
+			zc_warning_prev_line(
+					"Duplicate key found: key%d\n", key);
+		else
+			zc_error_prev_line(
+					"Duplicate key found: key%d\n", key);
+	}
+	/* Checks when a mandatory key is present */
+	if (!paramkeys[SVCB_KEY_MANDATORY])
+		return;
+
+	size = rdata_atom_size(parser->current_rr.rdatas[2]);
+	mandatory_values = (void*)rdata_atom_data(parser->current_rr.rdatas[2]);
+	mandatory_values += 2; /* skip the key type and length */
+
+	if (size % 2)
+		zc_error_prev_line("mandatory rdata must be a multiple of shorts");
+		
+	else for (i = 0; i < (size - 4)/2; i++) {
+		key = ntohs(mandatory_values[i]);
+
+		if (paramkeys[key])
+			; /* pass */
+
+		else if (key < SVCPARAMKEY_COUNT) {
+			if(zone_is_slave(parser->current_zone->opts))
+				zc_warning_prev_line("mandatory SvcParamKey: %s is missing "
+						     "the record\n", svcparamkey_strs[key]);
+			else
+				zc_error_prev_line("mandatory SvcParamKey: %s is missing "
+						   "the record\n", svcparamkey_strs[key]);
+		} else {
+			if(zone_is_slave(parser->current_zone->opts))
+				zc_warning_prev_line("mandatory SvcParamKey: key%d is missing "
+						     "the record\n", key);
+			else
+				zc_error_prev_line("mandatory SvcParamKey: key%d is missing "
+						   "the record\n", key);
+		}
+
+		/* In draft-ietf-dnsop-svcb-https-04 Section 8
+		 * automatically mandatory MUST NOT appear in its own value-list
+		 */
+		if (key == SVCB_KEY_MANDATORY) {
+			if(zone_is_slave(parser->current_zone->opts))
+				zc_warning_prev_line("mandatory MUST not be included"
+						     " as mandatory parameter");
+			else
+				zc_error_prev_line("mandatory MUST not be included"
+						   " as mandatory parameter");
+		}
+		if (key == prev_key) {
+			if(zone_is_slave(parser->current_zone->opts))
+				zc_warning_prev_line("Keys inSvcParam mandatory "
```

Comment:
Fix in your, otherwise very nice, list of fix commits.

```suggestion
				zc_warning_prev_line("Keys in SvcParam mandatory "
```